src/account/account.entity.ts

```diff
@@ -174,28 +104,65 @@ export class Account extends BaseEntity {
             }
         }
 
-        return new URL(
-            `.ghost/activitypub/${type}/${post.uuid}`,
-            `${Account.protocol}://${this.site!.host}`,
-        );
+        return new URL(`.ghost/activitypub/${type}/${post.uuid}`, this.apId);
     }
 
-    private static protocol: 'http' | 'https' =
-        process.env.NODE_ENV === 'testing' ? 'http' : 'https';
-
-    static createFromData(data: AccountData) {
-        return new Account(
-            data.id,
-            data.uuid,
-            data.username,
-            data.name,
-            data.bio,
-            data.avatarUrl,
-            data.bannerImageUrl,
-            data.site,
-            data.apId,
-            data.url,
-            data.apFollowers,
-        );
+    updateProfile(params: ProfileUpdateParams) {
+        type P = ProfileUpdateParams;
+        const get = <K extends keyof P>(prop: K): P[K] =>
+            params[prop] === undefined ? this[prop] : params[prop];
+
+        return AccountEntity.create({
+            ...this,
+            username: get('username'),
+            name: get('name'),
+            bio: get('bio'),
+            avatarUrl: get('avatarUrl'),
+            bannerImageUrl: get('bannerImageUrl'),
+        });
     }
 }
+
+type ProfileUpdateParams = {
+    name?: string | null;
+    bio?: string | null;
+    username?: string;
+    avatarUrl?: URL | null;
+    bannerImageUrl?: URL | null;
+};
+
+/**
+ * Internal accounts require a `host` so we can calculate the ActivityPub URLs
+ */
+type InternalAccountDraftData = {
+    isInternal: true;
+    host: URL;
+    username: string;
+    name: string;
+    bio: string | null;
+    url: URL | null;
+    avatarUrl: URL | null;
+    bannerImageUrl: URL | null;
+};
+
+/**
+ * External accounts require the ActivityPub URLs to be passed in
+ */
+type ExternalAccountDraftData = {
+    isInternal: false;
+    username: string;
+    name: string;
+    bio: string | null;
+    url: URL | null;
+    avatarUrl: URL | null;
+    bannerImageUrl: URL | null;
+    apId: URL;
+    apFollowers: URL | null;
+};
+
+type AccountDraftData = InternalAccountDraftData | ExternalAccountDraftData;
+
+type Data<T> = {
+    // biome-ignore lint/suspicious/noExplicitAny: These are not publicly usable instances of any
```

Comment:
This is legit - the `any` here is just to match functions, which we strip out - it doesn't leak into our code

Reply:
Maybe update the message to include this?